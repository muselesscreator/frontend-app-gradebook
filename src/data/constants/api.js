import { configuration } from 'config';

export const baseUrl = `${configuration.LMS_BASE_URL}/api`;

/**
 * bulkGradesUrlByCourseAndRow(courseId, rowId)
 * returns the bulkGrades url with the given courseId and rowId.
 * @param {string} courseId - course identifier
 * @param {string} rowId - row/error identifier
 * @return {string} - bulk grades fetch url
 */
export const bulkGradesUrlByCourseAndRow = (courseId, rowId) => (
  `${baseUrl}/bulkGrades/course/${courseId}/?error_id=${rowId}`
);
