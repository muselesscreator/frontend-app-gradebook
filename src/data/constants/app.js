import { StrictDict } from 'utils';

export const modalFieldKeys = StrictDict({
  adjustedGradePossible: 'adjustedGradePossible',
  adjustedGradeValue: 'adjustedGradeValue',
  assignmentName: 'assignmentName',
  reasonForChange: 'reasonForChange',
  todaysDate: 'todaysDate',
  updateModuleId: 'updateModuleId',
  updateUserId: 'updateUserId',
  updateUserName: 'updateUserName',
  open: 'open',
});

export const localFilterKeys = StrictDict({
  assignmentGradeMax: 'assignmentGradeMax',
  assignmentGradeMin: 'assignmentGradeMin',
  courseGradeMax: 'courseGradeMax',
  courseGradeMin: 'courseGradeMin',
});
