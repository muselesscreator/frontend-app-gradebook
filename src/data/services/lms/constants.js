import { StrictDict } from 'utils';

export const pageSize = 25;
export const historyRecordLimit = 5;

export const paramKeys = StrictDict({
  cohortId: 'cohort_id',
  pageSize: 'page_size',
  userContains: 'user_contains',
  enrollmentMode: 'enrollment_mode',
  assignment: 'assignment',
  assignmentGradeMin: 'assignmentGradeMin',
  assignmentGradeMax: 'assignmentGradeMax',
  courseGradeMin: 'courseGradeMin',
  courseGradeMax: 'courseGradeMax',
  excludedCourseRoles: 'excluded_course_roles',
});
