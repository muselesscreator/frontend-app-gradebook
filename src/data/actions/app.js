import { StrictDict } from 'utils';
import { modalFieldKeys, localFilterKeys } from 'data/constants/app';
import { createActionFactory } from './utils';

export const dataKey = 'app';
const createAction = createActionFactory(dataKey);

const closeModal = createAction('closeModal');
/**
 * setCourseId(courseId)
 * loads courseID into local app state
 * @param {string} courseId - course ID from app context
 */
const setCourseId = createAction('setCourseId');
/**
  * setModalStateFromTable({ userEntry, subsection })
  * sets modalState
  * */
const setModalStateFromTable = createAction('setModalStateFromTable');
/**
 * setSearchValue(searchValue)
 * sets searchValue in local state
 * @param {string} searchValue - search input text
 */
const setSearchValue = createAction('setSearchValue');

const filterReducer = (payload) => (obj, key) => (
  (payload[key] !== undefined) ? { ...obj, [key]: payload[key] } : obj
);

/**
 * setLocalFilter(filterObject)
 * sets a number of localFilterKey fields on localFilters
 * @param {object} filterObject - an object of filter values
 */
const setLocalFilter = createAction('setLocalFilter', (filterObject) => ({
  payload: Object.keys(localFilterKeys).reduce(filterReducer(filterObject), {}),
}));

/**
 * setModalState(modalState)
 * sets a number of modalFieldKey fields on modalState
 * @param {object} modalState - an object of modal state values
 */
const setModalState = createAction('setModalState', (modalState) => ({
  payload: Object.keys(modalFieldKeys).reduce(filterReducer(modalState), {}),
}));

export default StrictDict({
  closeModal,
  setCourseId,
  setModalState,
  setSearchValue,
  setLocalFilter,
  setModalStateFromTable,
});
