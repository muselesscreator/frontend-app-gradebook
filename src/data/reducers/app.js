import initialFilters from '../constants/filters';
import { formatDateForDisplay } from '../actions/utils';
import actions from '../actions/app';
import filterActions from '../actions/filters';

const initialState = {
  courseId: '',
  filters: {
    assignmentGradeMax: initialFilters.assignmentGradeMax,
    assignmentGradeMin: initialFilters.assignmentGradeMin,
    courseGradeMax: initialFilters.courseGradeMax,
    courseGradeMin: initialFilters.courseGradeMin,
  },
  modalState: {
    open: false,
    adjustedGradePossible: '',
    adjustedGradeValue: 0,
    assignmentName: '',
    reasonForChange: '',
    todaysDate: '',
    updateModuleId: null,
    updateUserId: null,
    updateUserName: null,
  },
  searchValue: '',
};

const app = (state = initialState, { type, payload }) => {
  switch (type) {
    case actions.closeModal.toString():
      return { ...state, modalState: { ...initialState.modalState } };
    case actions.setCourseId.toString():
      return { ...state, courseId: payload };
    case actions.setLocalFilter.toString():
      return {
        ...state,
        filters: { ...state.filters, ...payload },
      };
    case actions.setModalState.toString():
      return {
        ...state,
        modalState: { ...state.modalState, ...payload },
      };
    case actions.setModalStateFromTable.toString(): {
      const { subsection, userEntry } = payload;
      return {
        ...state,
        modalState: {
          adjustedGradeValue: '',
          open: true,
          reasonForChange: '',
          todaysDate: formatDateForDisplay(new Date()),

          adjustedGradePossible: subsection.attempted ? subsection.score_possible : '',
          assignmentName: `${subsection.subsection_name}`,
          updateModuleId: subsection.module_id,
          updateUserId: userEntry.user_id,
          updateUserName: userEntry.username,
        },
      };
    }
    case actions.setSearchValue.toString():
      return { ...state, searchValue: payload };
    // initialize the filter fields that are locally stored
    case filterActions.initialize.toString():
      return {
        ...state,
        filters: {
          assignmentGradeMax: payload.assignmentGradeMax,
          assignmentGradeMin: payload.assignmentGradeMin,
          courseGradeMax: payload.courseGradeMax,
          courseGradeMin: payload.courseGradeMin,
        },
      };
    // Reset only the filter fields that are stored locally
    case filterActions.reset.toString(): {
      return payload.filter(
        (filterName) => initialState.filters[filterName] !== undefined,
      ).reduce((obj, filterName) => ({
        ...obj,
        filters: {
          ...obj.filters,
          [filterName]: initialFilters[filterName],
        },
      }), { ...state });
    }
    default:
      return state;
  }
};

export { initialState };
export default app;
