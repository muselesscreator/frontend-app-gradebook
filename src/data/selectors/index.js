/* eslint-disable import/no-named-as-default-member, import/no-self-import */
import { StrictDict } from 'utils';
import LmsApiService from 'data/services/LmsApiService';

import * as module from '.';
import app from './app';
import assignmentTypes from './assignmentTypes';
import cohorts from './cohorts';
import filters from './filters';
import grades from './grades';
import roles from './roles';
import special from './special';
import tracks from './tracks';

/**
 * lmsApiServiceArgs(state)
 * Returns common lms api service request args.
 * @param {object} state - redux state
 * @return {object} lms api query params object
 */
export const lmsApiServiceArgs = (state) => ({
  cohort: cohorts.getCohortNameById(state, filters.cohort(state)),
  assignment: filters.selectedAssignmentId(state),
  assignmentType: filters.assignmentType(state),
  assignmentGradeMin: grades.formatMinAssignmentGrade(
    filters.assignmentGradeMin(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  assignmentGradeMax: grades.formatMaxAssignmentGrade(
    filters.assignmentGradeMax(state),
    { assignmentId: filters.selectedAssignmentId(state) },
  ),
  courseGradeMin: grades.formatMinCourseGrade(filters.courseGradeMin(state)),
  courseGradeMax: grades.formatMaxCourseGrade(filters.courseGradeMax(state)),
});

/**
 * gradeExportUrl(state, options)
 * Returns the output of getGradeExportCsvUrl, applying the current includeCourseRoleMembers
 * filter.
 * @param {object} state - redux state
 * @param {object} options - options object of the form ({ courseId })
 * @return {string} - generated grade export url
 */
export const gradeExportUrl = (state, { courseId }) => (
  LmsApiService.getGradeExportCsvUrl(courseId, {
    ...module.lmsApiServiceArgs(state),
    excludeCourseRoles: filters.includeCourseRoleMembers(state) ? '' : 'all',
  })
);

/**
 * interventionExportUrl(state, options)
 * Returns the output of getInterventionExportUrl.
 * @param {object} state - redux state
 * @param {object} options - options object of the form ({ courseId })
 * @return {string} - generated intervention export url
 */
export const interventionExportUrl = (state, { courseId }) => (
  LmsApiService.getInterventionExportCsvUrl(
    courseId,
    module.lmsApiServiceArgs(state),
  )
);

/**
 * getHeadings(state)
 * Returns the table headings given the current assignmentType and assignmentLabel filters.
 * @param {object} state - redux state
 * @return {string[]} - array of table headings
 */
export const getHeadings = (state) => grades.headingMapper(
  filters.assignmentType(state) || 'All',
  filters.selectedAssignmentLabel(state) || 'All',
)(grades.getExampleSectionBreakdown(state));

/**
 * showBulkManagement(state, options)
 * Returns true iff the user has special access or bulk management is configured to be available
 * and the course has a masters track.
 * @param {object} state - redux state
 * @param {object} options - options object of the form ({ courseId })
 * @return {bool} - should show bulk management controls?
 */
export const showBulkManagement = (state, { courseId }) => (
  special.hasSpecialBulkManagementAccess(courseId)
  || (tracks.stateHasMastersTrack(state) && state.config.bulkManagementAvailable)
);

/**
 * shouldShowSpinner(state)
 * Returns true iff the user can view the gradebook and grades.showSpinner is true.
 * @param {object} state - redux state
 * @return {bool} - should show spinner?
 */
export const shouldShowSpinner = (state) => (
  roles.canUserViewGradebook(state)
  && grades.showSpinner(state)
);

/**
 * showBulkManagement(state, options)
 * Returns true iff the user has special access or bulk management is configured to be available
 * and the course has a masters track.
 * @param {object} state - redux state
 * @return {bool} - should show bulk management controls?
 */
export const showBulkManagement = (state) => (
  special.hasSpecialBulkManagementAccess(app.courseId(state))
  || (tracks.stateHasMastersTrack(state) && state.config.bulkManagementAvailable)
);

export default StrictDict({
  root: StrictDict({
    getHeadings,
    gradeExportUrl,
    interventionExportUrl,
    shouldShowSpinner,
  }),
  app,
  assignmentTypes,
  cohorts,
  filters,
  grades,
  roles,
  special,
  tracks,
});
