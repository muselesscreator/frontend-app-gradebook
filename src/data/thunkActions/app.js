import { StrictDict } from 'utils';

import actions from 'data/actions';
import { fetchGradeOverrideHistory } from './grades';
import { fetchRoles } from './roles';

export const setModalStateFromTable = ({ userEntry, subsection }) => (
  (dispatch) => {
    dispatch(fetchGradeOverrideHistory(subsection.module_id, userEntry.user_id));
    dispatch(actions.app.setModalStateFromTable({ subsection, userEntry }));
  }
);

export const initialize = (courseId, urlQuery) => (dispatch) => {
  dispatch(actions.app.setCourseId(courseId));
  dispatch(actions.filters.initialize(urlQuery));
  dispatch(fetchRoles());
};

export default StrictDict({
  initialize,
  setModalStateFromTable,
});
