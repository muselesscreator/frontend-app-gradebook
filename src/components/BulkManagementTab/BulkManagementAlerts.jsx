/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { Alert } from '@edx/paragon';

import * as appConstants from 'data/constants/app';
import selectors from 'data/selectors';

const { messages: { BulkManagementTab: messages } } = appConstants;

export const BulkManagementAlerts = ({ bulkImportError, uploadSuccess }) => (
  <>
    <Alert
      variant="danger"
      show={!!bulkImportError}
      dismissible={false}
    >
      {bulkImportError}
    </Alert>
    <Alert
      variant="success"
      show={uploadSuccess}
      dismissible={false}
    >
      {messages.successDialog}
    </Alert>
  </>
);

BulkManagementAlerts.defaultProps = {
  bulkImportError: '',
  uploadSuccess: false,
};

BulkManagementAlerts.propTypes = {
  // redux
  bulkImportError: PropTypes.string,
  uploadSuccess: PropTypes.bool,
};

export const mapStateToProps = (state) => ({
  bulkImportError: selectors.grades.bulkImportError(state),
  uploadSuccess: selectors.grades.uploadSuccess(state),
});

export default connect(mapStateToProps)(BulkManagementAlerts);
