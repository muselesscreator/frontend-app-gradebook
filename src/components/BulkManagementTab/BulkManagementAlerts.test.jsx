import React from 'react';
import { shallow } from 'enzyme';
import { Alert } from '@edx/paragon';

import selectors from 'data/selectors';
import * as appConstants from 'data/constants/app';

import { BulkManagementAlerts, mapStateToProps } from './BulkManagementAlerts';

jest.mock('@edx/paragon', () => ({
  Alert: () => 'Alert',
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    grades: {
      bulkImportError: (state) => ({ bulkImportError: state }),
      uploadSuccess: (state) => ({ uploadSuccess: state }),
    },
  },
}));

const errorMessage = 'Oh noooooo';

describe('BulkManagementAlerts', () => {
  describe('component', () => {
    let el;
    describe('no errer, no upload success', () => {
      beforeEach(() => {
        el = shallow(<BulkManagementAlerts />);
      });
      test('snapshot - bulkImportError closed, success closed', () => {
        expect(el).toMatchSnapshot();
      });
      test('closed danger alert', () => {
        expect(el.childAt(0).is(Alert)).toEqual(true);
        expect(el.childAt(0).props().show).toEqual(false);
        expect(el.childAt(0).props().variant).toEqual('danger');
      });
      test('closed success alert', () => {
        expect(el.childAt(1).is(Alert)).toEqual(true);
        expect(el.childAt(1).props().show).toEqual(false);
        expect(el.childAt(1).props().variant).toEqual('success');
      });
    });
    describe('no errer, no upload success', () => {
      beforeEach(() => {
        el = shallow(<BulkManagementAlerts uploadSuccess bulkImportError={errorMessage} />);
      });
      const assertions = [
        'danger alert open with bulkImportError',
        'success alert open with messages.successDialog',
      ];
      test(`snapshot - ${assertions.join(', ')}`, () => {
        expect(el).toMatchSnapshot();
      });
      test('open danger alert with bulkImportError content', () => {
        expect(el.childAt(0).is(Alert)).toEqual(true);
        expect(el.childAt(0).children().text()).toEqual(errorMessage);
        expect(el.childAt(0).props().show).toEqual(true);
      });
      test('open success alert with messages.successDialog content', () => {
        expect(el.childAt(1).is(Alert)).toEqual(true);
        expect(el.childAt(1).children().text()).toEqual(
          appConstants.messages.BulkManagementTab.successDialog,
        );
        expect(el.childAt(1).props().show).toEqual(true);
      });
    });
  });
  describe('mapStateToProps', () => {
    let mapped;
    const testState = { a: 'puppy', named: 'Ember' };
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('bulkImportError from grades.bulkImportError', () => {
      expect(mapped.bulkImportError).toEqual(selectors.grades.bulkImportError(testState));
    });
    test('uploadSuccess from grades.uploadSuccess', () => {
      expect(mapped.uploadSuccess).toEqual(selectors.grades.uploadSuccess(testState));
    });
  });
});
