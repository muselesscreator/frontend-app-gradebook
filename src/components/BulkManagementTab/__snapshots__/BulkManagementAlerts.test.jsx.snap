// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkManagementAlerts component no errer, no upload success snapshot - bulkImportError closed, success closed 1`] = `
<Fragment>
  <Alert
    dismissible={false}
    show={false}
    variant="danger"
  />
  <Alert
    dismissible={false}
    show={false}
    variant="success"
  >
    CSV processing. File uploads may take several minutes to complete.
  </Alert>
</Fragment>
`;

exports[`BulkManagementAlerts component no errer, no upload success snapshot - danger alert open with bulkImportError, success alert open with messages.successDialog 1`] = `
<Fragment>
  <Alert
    dismissible={false}
    show={true}
    variant="danger"
  >
    Oh noooooo
  </Alert>
  <Alert
    dismissible={false}
    show={true}
    variant="success"
  >
    CSV processing. File uploads may take several minutes to complete.
  </Alert>
</Fragment>
`;
